tools/clang/test/HLSL/workgraph/dispatchgrid_diags.hlsl

```diff
@@ -55,41 +62,92 @@ void node06()
 [NodeMaxDispatchGrid(65535, 1, 1)]
 [NumThreads(32, 1, 1)]
-void node11()
+void node11(DispatchNodeInputRecord<MyStruct> input)
 { }
 
 [Shader("node")]
 [NodeLaunch("broadcasting")]
 [NodeMaxDispatchGrid(1, 65536, 1)]    // expected-error {{'NodeMaxDispatchGrid' Y component value must be between 1 and 65,535 (2^16-1) inclusive}}
 [NumThreads(32, 1, 1)]
-void node12()
+void node12(DispatchNodeInputRecord<MyStruct> input)
 { }
 
 [Shader("node")]
 [NodeLaunch("broadcasting")]
 [NodeMaxDispatchGrid(1, 1, 0)]        // expected-error {{'NodeMaxDispatchGrid' Z component value must be between 1 and 65,535 (2^16-1) inclusive}}
 [NumThreads(32, 1, 1)]
-void node13()
+void node13(DispatchNodeInputRecord<MyStruct> input)
 { }
 
 [Shader("node")]
 [NodeLaunch("broadcasting")]
 [NodeMaxDispatchGrid(1, y, z)]        // expected-error {{'NodeMaxDispatchGrid' Z component value must be between 1 and 65,535 (2^16-1) inclusive}}
 [NumThreads(32, 1, 1)]
-void node14()
+void node14(DispatchNodeInputRecord<MyStruct> input)
 { }
 
 [Shader("node")]
 [NodeLaunch("broadcasting")]
 [NodeMaxDispatchGrid(1, 65535, 257)]  // expected-error {{'NodeMaxDispatchGrid' X * Y * Z product may not exceed 16,777,215 (2^24-1)}}
 [NumThreads(32, 1, 1)]
-void node15()
+void node15(DispatchNodeInputRecord<MyStruct> input)
 { }
 
 [Shader("node")]
 [NodeLaunch("broadcasting")]
 [NodeMaxDispatchGrid(256, 8, 8)]      // expected-warning {{attribute 'NodeMaxDispatchGrid' is already applied}}
 [NodeMaxDispatchGrid(64, 1, 1)]
 [NumThreads(32, 1, 1)]
-void node16()
+void node16(DispatchNodeInputRecord<MyStruct> input)
+{ }
+
+[Shader("node")]
+[NodeMaxDispatchGrid(3, 1, 1)]
+[NumThreads(17, 1, 1)]
+void node17()               // expected-error {{Broadcasting node shader 'node17' with NodeMaxDispatchGrid attribute must declare an input record containing a field with SV_DispatchGrid semantic}}
+{ }
+
+struct MyStruct2 {
+    float3 b;
+};
+
+[Shader("node")]
+[NodeLaunch("broadcasting")]
+[NodeMaxDispatchGrid(256, 8, 8)]
+[NumThreads(32, 1, 1)]
+void node18(DispatchNodeInputRecord<MyStruct2> input)  // expected-error {{Broadcasting node shader 'node18' with NodeMaxDispatchGrid attribute must declare an input record containing a field with SV_DispatchGrid semantic}}
+{ }
+
+struct MyStruct3 {
+    uint3 grid : SV_DispatchGrid;  // expected-note {{other SV_DispatchGrid defined here}}
+    float3 b;
+    int3 grid2 : SV_DispatchGrid;  // expected-error {{a field with SV_DispatchGrid has already been specified}}
+};
+
+[Shader("node")]
+[NodeLaunch("broadcasting")]
+[NodeMaxDispatchGrid(256, 8, 8)]
+[NumThreads(32, 1, 1)]
+void node19(DispatchNodeInputRecord<MyStruct3> input)
+{ }
+
+struct A {
+    float a;
+    uint3 grid : SV_DispatchGrid;   // expected-error {{a field with SV_DispatchGrid has already been specified}}
+};
+
+struct B : A {
+    float b;
+};
+
+struct C : B {
+    float b;
+    int3 grid2 : SV_DispatchGrid;   // expected-note {{other SV_DispatchGrid defined here}}
```

Comment:
Oh, I see that derived case does in fact work due to the way `InputRecordDecl->fields()` must iterate base class fields as well.  But I doubt the nested case does.  And what about templates?  Also, the locations of the error and note seems reversed from what I would expect.

Here's a fun one to try:
```hlsl
template<typename T, int C> struct DispatchGridField {
  vector<T, C> grid : SV_DispatchGrid;  // expected-note {{other SV_DispatchGrid defined here}}
};
template<typename T, int C> struct UseDispatchGrid {
  DispatchGridField<T, C> dg;
};
template <typename T, int C> struct Record : UseDispatchGrid<T, C> {
  uint grid2 : SV_DispatchGrid;  // expected-error {{a field with SV_DispatchGrid has already been specified}}
};

[Shader("node")]
[NodeLaunch("broadcasting")]
[NodeMaxDispatchGrid(32, 16, 1)]
[NumThreads(32, 1, 1)]
void node1(DispatchNodeInputRecord< Record<uint, 2> > input) {}
```

Reply:
Templates don't work at all :( 
Issue filed: #5842.